lib/spack/docs/build_systems/cmakepackage.rst

```diff
@@ -0,0 +1,258 @@
+installation directory, and set the build type.
+
+The rest of the flags are the ones you should consider adding to your
+package. They often include flags to enable/disable support for certain
+features and locate specific dependencies. One thing you'll notice that
+makes CMake different from Autotools is that CMake has an understanding
+of build flag hierarchy. That is, certain flags will not display unless
+their parent flag has been selected. For example, flags to specify the
+``lib`` and ``include`` directories for a package might not appear
+unless CMake found the dependency it was looking for. You may need to
+manually specify certain flags to explore the full depth of supported
+build flags, or check the ``CMakeLists.txt`` yourself.
+
+^^^^^^^^^^^^^^^^^^^^^
+Adding flags to cmake
+^^^^^^^^^^^^^^^^^^^^^
+
+To add additional flags to the ``cmake`` call, simply override the
+``cmake_args`` function:
+
+.. code-block:: python
+
+   def cmake_args(self):
+       args = []
+
+       if '+hdf5' in self.spec:
+           args.append('-DDETECT_HDF5=ON')
+       else:
+           args.append('-DDETECT_HDF5=OFF')
+
+       return args
+
+
+^^^^^^^^^^
+Generators
+^^^^^^^^^^
+
+CMake actually supports multiple generators, not just Makefiles.
+Another common generator is Ninja. To switch to the Ninja
+generator, simply add:
+
+.. code-block:: python
+
+   generator = 'Ninja'
+
+
+``CMakePackage`` defaults to "Unix Makefiles". If you switch to the
+Ninja generator, make sure to add:
+
+.. code-block:: python
+
+   depends_on('ninja', type='build')
+
+to the package as well. Aside from that, you shouldn't need to do
+anything else. Spack will automatically detect that you are using
+Ninja and run:
+
+.. code-block:: console
+
+   $ cmake .. -G Ninja
+   $ ninja
+   $ ninja install
+
+Spack currently only supports "Unix Makefiles" and "Ninja" as valid
+generators, but it should be simple to add support for alternative
+generators. For more information on CMake generators, see:
+https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html
+
+^^^^^^^^^^^^^^^^
+CMAKE_BUILD_TYPE
+^^^^^^^^^^^^^^^^
+
+Every CMake-based package accepts a ``-DCMAKE_BUILD_TYPE`` flag to
+dictate which level of optimization to use. In order to ensure
+uniformity across packages, the ``CMakePackage`` base class adds
+a variant to control this:
+
+.. code-block:: python
+
+   variant('build_type', default='RelWithDebInfo',
+           description='CMake build type',
+           values=('Debug', 'Release', 'RelWithDebInfo', 'MinSizeRel'))
+
+However, not every CMake package accepts all four of these options.
+Grep the ``CMakeLists.txt`` file to see if the default values are
+missing or replaced. For example, the ``dealii`` package overrides
+the default variant with:
+
+.. code-block:: python
+
+   variant('build_type', default='DebugRelease',
+           description='The build type to build',
+           values=('Debug', 'Release', 'DebugRelease'))
```

Comment:
Should dealii's `DebugRelease` be changed to `RelWithDebInfo`?  Why is `dealii` nonstandard anyway?  Can we remove this?

Reply:
Unfortunately, `dealii` isn't the only one:
```console
$ grep 'variant(.*build_type' */package.py
benchmark/package.py:    variant('build_type', default='RelWithDebInfo',
cbtf-argonavis/package.py:    variant('build_type', default='None', values=('None'),
cbtf-krell/package.py:    variant('build_type', default='None', values=('None'),
cbtf-lanl/package.py:    variant('build_type', default='None', values=('None'),
cbtf/package.py:    variant('build_type', default='None', values=('None'),
cgal/package.py:    variant('build_type', default='Release',
dealii/package.py:    variant('build_type', default='DebugRelease',
eccodes/package.py:    variant('build_type', default='RelWithDebInfo',
eigen/package.py:    variant('build_type', default='RelWithDebInfo',
elemental/package.py:    variant('build_type', default='Release',
fenics/package.py:    variant('build_type', default='RelWithDebInfo',
grib-api/package.py:    variant('build_type', default='RelWithDebInfo',
gromacs/package.py:    variant('build_type', default='RelWithDebInfo',
hydrogen/package.py:    variant('build_type', default='Release',
jsoncpp/package.py:    variant('build_type', default='RelWithDebInfo',
krims/package.py:    variant("build_type", default="DebugRelease",
lazyten/package.py:    variant("build_type", default="DebugRelease",
lbann/package.py:    variant('build_type', default='Release',
libemos/package.py:    variant('build_type', default='RelWithDebInfo',
llvm-openmp-ompt/package.py:    variant('build_type', default='Release',
llvm/package.py:    variant('build_type', default='Release',
magics/package.py:    variant('build_type', default='RelWithDebInfo',
mbedtls/package.py:    variant('build_type', default='Release',
metis/package.py:    variant('build_type', default='Release',
opencoarrays/package.py:    variant('build_type', default='RelWithDebInfo',
openspeedshop/package.py:    variant('build_type', default='None', values=('None'),
psi4/package.py:    variant('build_type', default='Release',
relion/package.py:    variant('build_type', default='RelWithDebInfo',
rr/package.py:    variant('build_type', default='Release',
symengine/package.py:    variant('build_type', default='Release',
tethex/package.py:    variant('build_type', default='Release',
typhonio/package.py:    variant('build_type', default='Release', description='The build type to build',
vc/package.py:    variant('build_type', default='RelWithDebInfo',
```
Maybe @davydden can comment on this.